Modules/_localemodule.c

```diff
@@ -602,14 +653,27 @@ _locale_nl_langinfo_impl(PyObject *module, int item)
     /* Check whether this is a supported constant. GNU libc sometimes
        returns numeric values in the char* return value, which would
        crash PyUnicode_FromString.  */
-    for (i = 0; langinfo_constants[i].name; i++)
+    for (i = 0; langinfo_constants[i].name; i++) {
         if (langinfo_constants[i].value == item) {
             /* Check NULL as a workaround for GNU libc's returning NULL
                instead of an empty string for nl_langinfo(ERA).  */
             const char *result = nl_langinfo(item);
             result = result != NULL ? result : "";
-            return PyUnicode_DecodeLocale(result, NULL);
+            if (langinfo_constants[i].category == LC_CTYPE || is_all_ascii(result)) {
+                return PyUnicode_DecodeLocale(result, NULL);
+            }
+
+            char *oldloc;
+            if (change_locale(langinfo_constants[i].category, &oldloc) < 0) {
+                return NULL;
+            }
+            result = nl_langinfo(item);
```

Comment:
Is it needed to call nl_langinfo() twice? If yes, would you mind to add a short comment explaining why?

Reply:
The result is not valid after calling `setlocale()`.

Reply:
On other hand, this may be overprotection. [The POSIX documentation](https://pubs.opengroup.org/onlinepubs/009696699/functions/nl_langinfo.html) says: _"Calls to [setlocale()](https://pubs.opengroup.org/onlinepubs/009696699/functions/setlocale.html) with a category corresponding to the category of item (see [<langinfo.h>](https://pubs.opengroup.org/onlinepubs/009696699/basedefs/langinfo.h.html)), or to the category LC_ALL , may overwrite the array pointed to by the return value."_

Reply:
I removed the comment and simplified the code. It may become more complicated with #124969.